### ISSUE TITLE
[css-cascade-6] `&` matching inside the `@scope`, and its interaction with `:scope`

### ISSUE THREAD

**Issue by kizu** [w3c/csswg-drafts#9740]
In some way, this is a follow-up to the https://github.com/w3c/csswg-drafts/issues/9739, but I felt that this is a different case that could be discussed separately.

While testing different ways scopes interact when being nested, and how `&` and `:scope` behave inside, I found one place that should either be clarified in the specs, or maybe changed.

I'm talking about two parts in particular.

From [2.5.1. Effects of `@scope`](https://drafts.csswg.org/css-cascade-6/#scope-effects)

> The [`&`](https://drafts.csswg.org/css-nesting-1/#selectordef-) selector is defined to represent the selector representing the scoping root (the [`<scope-start>`](https://drafts.csswg.org/css-cascade-6/#typedef-scope-start) selector), or else `:scope` if no selector was specified.

From [2.5.3. Scoped Style Rules](https://drafts.csswg.org/css-cascade-6/#scoped-rules)

> Any selector in the [`<relative-selector-list>`](https://drafts.csswg.org/selectors-4/#typedef-relative-selector-list) that does not start with a [combinator](https://drafts.csswg.org/selectors-4/#selector-combinator) but does [contain the nesting selector](https://drafts.csswg.org/css-nesting-1/#contain-the-nesting-selector) or the [`:scope`](https://drafts.csswg.org/selectors-4/#scope-pseudo) selector, is interpreted as a non-[relative selector](https://drafts.csswg.org/selectors-4/#relative-selector)

The short version of my question: Should `&` represent **only** the scoping root selector, **or** both the selector and `:scope`?

The reasoning behind my question: right now, there is a difference when we use `&` or `&:scope` inside a scope.

When we have only `&`, we, essentially, lose the `:scope` completely, making it so the `&` will match _any_ similar node, not necessarily if it is *the* scoping root node.

Here is a CodePen demonstrating the current behavior: https://codepen.io/kizu/pen/GReRRxQ

So, the way it works now, the `&` and `:scope` are different in two ways:

1. `:scope` has a specificity of a pseudo-class, and `&` has a specificity of the scoping root selector.
2. The `:scope` matches the scoping root node, but `&` matches only the scoping root selector itself.

The current behavior can be potentially useful, but I'm struggling to come up with actual use cases. We can also always add an explicit `:scope` and write `&:scope` to get both effects (or `&:where(:scope)` if we'd want to only have the specificity of the scoping root selector).

However, there is a chance this difference could result in author confusion and mistakes, as it is very easy to assume that the specificity difference is _the only_ difference.

If this behavior is to stay, I think the specs should be more explicit in how they explain these differences.

Otherwise, I think it might be better to change the `&` to be the same as `:scope` — representing the scoping root **node**, but with a different specificity.

**Comment by andruud** [comment]
> Otherwise, I think it might be better to change the & to be the same as :scope — representing the scoping root node, but with a different specificity.

Hmm, are you sure it's not _more_ confusing to deviate from how `&` behaves in css-nesting?

**Comment by mirisuzanne** [comment]
I agree that there's potential for author confusion - but I don't think it's improved by making them more similar, while still not identical, and making `&` behave in special ways here.

**Comment by bramus** [comment]
> https://blog.kizu.dev/scope-selector-nuance/

I don’t see the problem there, and think the stuff you described there is WAI.

> PROPOSED: `&` in `@scope` behaves as `:where(:scope)`.

This feels weird to me, as it’s deviating from what `&` does in nesting.

---

One thing you could argue here instead, is whether using `@scope` should carve out a subtree based on `<scope-start>` and only consider that subtree in isolation when doing selector matching? Right now it doesn’t, but if it did then it would open up the door to possible performance gains for authors. It would also bypass the whole `.outer :scope .inner` situation.

**Comment by romainmenke** [comment]
> One thing you could argue here instead, is whether using `@scope` should carve out a subtree based on <scope-start> and only consider that subtree in isolation when doing selector matching? Right now it doesn’t, but if it did then it would open up the door to possible performance gains for authors. It would also bypass the whole `.outer :scope .inner` situation.

Does this make it impossible to have `.outer :scope .inner`  match anything?
Because that will take away a lot of utility, right?

Or do you suggest making the implicit case behave more like `.outer .inner:is(:scope *)`?

**Comment by andruud** [comment]
> One thing you could argue here instead, is whether using @scope should carve out a subtree based on <scope-start> and only consider that subtree in isolation when doing selector matching?

We already basically have this property with the current design. There's no point in looking at `.outer :scope .inner` at all, if the element you're currently considering isn't in scope.

**Comment by kizu** [comment]
> @bramus 
> > https://blog.kizu.dev/scope-selector-nuance/
> I don’t see the problem there, and think the stuff you described there is WAI.

I don't see a problem with it, I shared the link as a part of a question if selector can contain something that is outside of the scope. 

>> PROPOSED: `&` in `@scope` behaves as `:where(:scope)`.
> This feels weird to me, as it’s deviating from what `&` does in nesting.

But we're not exactly in the nesting case with the `@scope`: it is a new scenario, where we have some selector as a scoping root, _which contributes zero specificity_.

If we were to express

```css
@scope (.foo) {
    .bar {…}
}
```

as nesting equivalent right now (omitting the scoping nuances), it would be 

```css
:where(.foo) {
    .bar {…}
}
```

As the scoping root _does not_ contribute any specificity.

And the problem is this implicit “nesting”: having the specificity of this be different from 

```css
@scope (.foo) {
    & .bar {…}
}
```

is pretty weird, because right now the absence of `&` _does_ behave like an implicit `:where(:scope)` already. And if we want to have consistency with nesting, then having `.bar` and `& .bar` be fully equivalent will be less confusing. See https://codepen.io/kizu/pen/VYLJwzZ

So it would be great to make `.bar` and `& .bar` behave the same.

> One thing you could argue here instead, is whether using `@scope` should carve out a subtree based on `<scope-start>` and only consider that subtree in isolation when doing selector matching? Right now it doesn’t, but if it did then it would open up the door to possible performance gains for authors. It would also bypass the whole `.outer :scope .inner` situation.

I don't see the issue with the `.outer :scope .inner` specifically, and as @romainmenke I see the utility of this. If we'd want to have a way to clarify that some part of a selector should target _only_ something inside the scope, this is a different problem, likely covered by https://github.com/w3c/csswg-drafts/issues/9841 (where it is similar with `:has()` checking for something in the donut).

**Comment by mirisuzanne** [comment]
I wrote up a [short post explaining the decision](https://css.oddbird.net/scope/parent-selector/). @fantasai feel free to share that in whatever channels make sense.

**Comment by bramus** [comment]
> I wrote up a [short post explaining the decision](https://css.oddbird.net/scope/parent-selector/). [@fantasai](https://github.com/fantasai) feel free to share that in whatever channels make sense.

Ah, this was easier to follow than the minutes. I think I get it now, and can follow the reasoning.

**Comment by mirisuzanne** [comment]
@astearns I marked this as accepted and needs-edits, but I wasn't around for the 'revisit in one weeks time' - so not sure if that happened.

**Comment by astearns** [comment]
Bah, I missed it. The revisit will be first thing this week.
